src/shogun/classifier/LDA.cpp

```diff
@@ -95,45 +113,49 @@ bool CLDA::train_machine(CFeatures *data)
 			classidx_pos[num_pos++]=i;
 	}
 
-	w=SGVector<float64_t>(num_feat);
-	w.zero();
-	MatrixXd fmatrix=Map<MatrixXd>(feature_matrix.matrix, num_feat, num_vec);
-	VectorXd mean_neg(num_feat);
-	mean_neg=VectorXd::Zero(num_feat);
-	VectorXd mean_pos(num_feat);
-	mean_pos=VectorXd::Zero(num_feat);
+	SGVector<ST> w_st(num_feat);
+	w_st.zero();
+	typename SGMatrix<ST>::EigenMatrixXt fmatrix = typename SGMatrix<ST>::EigenMatrixXtMap(feature_matrix.matrix, num_feat, num_vec);
+	typename SGVector<ST>::EigenVectorXt mean_neg(num_feat);
+	mean_neg.setZero();
+	typename SGVector<ST>::EigenVectorXt mean_pos(num_feat);
+	mean_pos.setZero();
 
 	//mean neg
-	for(i=0; i<num_neg; i++)
-		mean_neg+=fmatrix.col(classidx_neg[i]);
-	mean_neg/=(float64_t)num_neg;
+	for(i=0; i<num_neg; i++){
+		mean_neg+= fmatrix.col(classidx_neg[i]);
+	}
+
+	mean_neg/= (ST) num_neg;
 
 	// get m(-ve) - mean(-ve)
 	for(i=0; i<num_neg; i++)
-		fmatrix.col(classidx_neg[i])-=mean_neg;
+		fmatrix.col(classidx_neg[i])-= mean_neg;
 
 	//mean pos
 	for(i=0; i<num_pos; i++)
-		mean_pos+=fmatrix.col(classidx_pos[i]);
-	mean_pos/=(float64_t)num_pos;
+		mean_pos += fmatrix.col(classidx_pos[i]);
+	mean_pos/= (ST) num_pos;
 
 	// get m(+ve) - mean(+ve)
 	for(i=0; i<num_pos; i++)
-		fmatrix.col(classidx_pos[i])-=mean_pos;
+		fmatrix.col(classidx_pos[i])-= mean_pos;
 
-	SGMatrix<float64_t>scatter_matrix(num_feat, num_feat);
-	Map<MatrixXd> scatter(scatter_matrix.matrix, num_feat, num_feat);
+	SGMatrix<ST>scatter_matrix(num_feat, num_feat);
+	typename SGMatrix<ST>::EigenMatrixXtMap scatter(scatter_matrix.matrix, num_feat, num_feat);
 
 	if (m_method == FLD_LDA || (m_method==AUTO_LDA && num_vec>num_feat))
 	{
 		// covariance matrix.
-		MatrixXd cov_mat(num_feat, num_feat);
+		typename SGMatrix<ST>::EigenMatrixXt cov_mat(num_feat, num_feat);
 		cov_mat=fmatrix*fmatrix.transpose();
 		scatter=cov_mat/(num_vec-1);
-		float64_t trace=scatter.trace();
-		double s=1.0-m_gamma;
-		scatter *=s;
-		scatter.diagonal()+=VectorXd::Constant(num_feat, trace*m_gamma/num_feat);
+		ST trace=scatter.trace();
```

Comment:
I'm a bit confused - ```scatter``` is an an eigen map, so ```trace()``` is eigen's trace method (```scatter_matrix``` is the SGMatrix that is wrapped by ```scatter```).